sound/soc/sof/topology.c

```diff
@@ -54,6 +54,81 @@ struct sof_widget_data {
 	struct snd_sof_control *control;
 };
 
+/*
+ * update the IPC struct based on the tuple array.
+ */
+int sof_update_ipc_object(struct snd_soc_component *scomp, void *object, enum sof_tokens token_id,
+			  int num_tuples, struct snd_sof_tuple *tuples,
+			  size_t object_size, int num_sets)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	const struct ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg_ipc_ops;
+	const struct sof_token_info *token_list = ipc_tplg_ops->token_list;
+	const struct sof_topology_token *tokens;
+	int i, j;
+
+	if (token_list[token_id].count < 0) {
+		dev_err(scomp->dev, "No tokens for token id: %d\n", token_id);
+		return -EINVAL;
+	}
+
+	tokens = token_list[token_id].tokens;
+	if (!tokens) {
+		dev_err(scomp->dev, "Invalid tokens for token id: %d\n", token_id);
+		return -EINVAL;
+	}
+
+	for (i = 0; i < token_list[token_id].count; i++) {
+		int offset = 0;
+		int num_tokens_matched = 0;
+
+		for (j = 0; j < num_tuples; j++) {
+			if (tokens[i].token == tuples[j].token) {
+				switch (tokens[i].type) {
+				case SND_SOC_TPLG_TUPLE_TYPE_WORD:
+				{
+					u32 *val = (u32 *)((u8 *)object + tokens[i].offset +
+							   offset);
+
+					*val = tuples[j].value.v;
+					break;
+				}
+				case SND_SOC_TPLG_TUPLE_TYPE_SHORT:
+				case SND_SOC_TPLG_TUPLE_TYPE_BOOL:
+				{
+					u16 *val = (u16 *)((u8 *)object + tokens[i].offset +
+							    offset);
+
+					*val = (u16)tuples[j].value.v;
+					break;
+				}
+				case SND_SOC_TPLG_TUPLE_TYPE_STRING:
+				{
+					/* all string tokens are converted to u32 values */
+					u32 *val = (u32 *)((u8 *)object + tokens[i].offset +
+							   offset);
+
+					*val = tokens[i].find_token((void *)tuples[j].value.s);
+				}
+				default:
+					break;
+				}
+
+				num_tokens_matched++;
+
+				/* found all required sets of current token */
+				if (!(num_tokens_matched % num_sets))
+					break;
+
+				/*  move to the next set */
+				offset += object_size;
+			}
+		}
+	}
+
+	return 0;
+}
```

Comment:
I am not clear on the code changes. We add this new ipc3-topology.c file with a ton of new code, but here there are only minor additions and no removal?

Reply:
removal is done in the next patch @plbossart. Ideally I could squash the two patches but that would make this commit a massive one.

Reply:
the next patch is not really a removal, there are tons of functional changes to topology.c

I am not saying I want a single patch, but more a split between moving code around in different files and changing code.

I know it's painful, but doing the 'right' thing upfront will make it easier for upstream/backports.

Reply:
@plbossart i've rearranged the commits and made the change in this patch a literal code move from topology.c to ipc3-topology.c along with the change in token parsing. Let me know if this reads better.